tao_compiler/mlir/disc/transforms/shape_utils.cc

```diff
@@ -997,6 +998,60 @@ LogicalResult ShapeAnalysis::applyMhloOpConstraint(Operation* op) {
       mapDimEqual(mn_values[i].first, mn_values[i].second, result,
                   i + lhs_batching_dims.size());
     }
+  } else if (auto einsum = dyn_cast<mhlo::EinsumOp>(op)) {
```

Comment:
I think the string tokenizer codes should be extracted to another method.

The `applyMhloOpConstraint` method is amazing long. Roman wasn't built in a day :).

Reply:
Good suggestion, I extract parseEinsumEquation into disc_util. This should be merged into hlo_ops.cc in future.